File: img2ascii.js
import { Jimp } from 'jimp';

async function convertToAscii(imagePath, targetWidth = 40) {
  const chars = ['@', '%', '#', '*', '+', '=', '-', ':', '.', ' '];
  
  try {
    const image = await Jimp.read(imagePath);
    
    // Resize image
    const w = image.bitmap.width;
    const h = image.bitmap.height;
    const aspect = h / w;
    // Monospace height to width ratio is roughly 1.6 - 2.0
    // So if we have width=40, the height should be proportionally smaller
    const targetHeight = Math.floor(targetWidth * aspect * 0.55);
    
    image.resize({ w: targetWidth, h: targetHeight });
    image.greyscale();
    
    let asciiArt = '';
    
    for (let y = 0; y < targetHeight; y++) {
      let row = '';
      for (let x = 0; x < targetWidth; x++) {
        // Get the greyscale value from 0 to 255
        // jimp returns hex rgba, we just need r since it's greyscale
        const hex = image.getPixelColor(x, y);
        const r = (hex >> 24) & 255;
        
        // Map 0-255 to 0-9
        let charIndex = Math.floor(r / 25.6);
        if (charIndex > 9) charIndex = 9;
        
        row += chars[charIndex];
      }
      asciiArt += row + '\n';
    }
    
    console.log(asciiArt);
  } catch (err) {
    console.error('Error:', err);
  }
}

convertToAscii('assets/me.png');

File: img2ascii.py
import sys
try:
    from PIL import Image
except ImportError:
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "pillow"])
    from PIL import Image

def image_to_ascii(image_path, width=45):
    # characters from dark to light
    chars = ["@", "%", "#", "*", "+", "=", "-", ":", ".", " "]
    
    try:
        img = Image.open(image_path).convert('L')
    except Exception as e:
        print(f"Error: {e}")
        return

    # Resize
    w, h = img.size
    aspect_ratio = h / float(w)
    # Monospace fonts in SVG usually have a height/width ratio of ~ 1.66
    new_height = int(width * aspect_ratio * 0.6)
    img = img.resize((width, new_height))
    
    pixels = img.getdata()
    ascii_str = ""
    for pixel in pixels:
        # 256 / len(chars) = 25.6
        index = int(pixel / 25.6)
        if index >= len(chars): index = len(chars)-1
        ascii_str += chars[index]
    
    # Format to string with newlines
    ascii_img = ""
    for i in range(0, len(ascii_str), width):
        ascii_img += ascii_str[i:i+width] + "\n"
        
    print(ascii_img)

if __name__ == "__main__":
    image_to_ascii("assets/me.png")
